### ISSUE TITLE
Rule proposal: require explicit {} generic type parameter

### ISSUE THREAD

**Issue by myitcv** [palantir/tslint#720]
This will also be dependent on #680 landing.

Consider:

``` typescript
class Wrapper<T> {
    val: T;
}

let x = new Wrapper();
```

The type of `x` here is `Wrapper<{}>` because:
- the compiler can't infer anything else
- I haven't supplied a type parameter

I would like there to be a lint error whenever the compiler falls back to `{}` as the inferred generic type parameter. 

If `{}` is what I intend then I can write:

``` typescript
let x = new Wrapper<{}>();
```

Thoughts?

**Comment by adidahiya** [comment]
nice! how about `no-default-type-param` as the rule name?

**Comment by adidahiya** [comment]
or perhaps something simpler, since it's enforcing the _existence_ of something: `generic-type-param`

**Comment by myitcv** [comment]
I'm useless at naming things :tada: 

`{}` is referred to as an [_empty object type parameter_](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#361-type-parameter-lists) in the spec.

So how about `explicit-empty-object-type-param`?

**Comment by andy-hanson** [comment]
Can this be closed by #1821?
